No longer picks unrepairable items for quick repair

zwparchman opened the pull request:
Previously damaged batteries and other things that could not be repaired
were selected with the quick repair action. This meant you couldn't hold
'r' to repair your vehicle if it had a damaged battery. The interface no
longer snaps to unrepairable parts when the repair key is pressed.

codemime commented on src/veh_interact.h:
```diff
@@ -152,6 +152,9 @@ class veh_interact
         int mostDamagedPart = -1;
+        /** Store the part that needs repair the most. This may not be 
+         * mostDamagedPart since not all parts can be repaired */
+        int mostRepairablePart = -1;
```
You could avoid introducing additional state by using a function (something like `get_most_repairable_part() const`).
`mostRepairablePart` is utilized in one place only and it doesn't affect overall performance at all.

Also, it would be nice to do the same for `mostDamagedPart`, but that's not mandatory here.

codemime commented:
If the most repairable part isn't the most damaged, the former should also be displayed via UI to prevent surprises.

zwparchman commented:
It's ready for round 2. 

I've updated the formatting and added a way to see most_repairable in the vehicle interface.

If the most damaged item is repairable this pull request should have no visible effect.
If the most damaged item isn't repairable the message "(can't repair)" is added to "Most Damaged:".
If the most damaged and most repairable are not the same item then a "Needs Repair:" message shows up below "Most Damaged" as well as the above effect.

codemime commented on src/veh_interact.h:
```diff
@@ -150,8 +150,13 @@ class veh_interact
+        /** Returns the index of the part that needs repair the most.
+         * This may not be mostDamagedPart since not all parts can be repaired
+         * If there are no damaged parts this returns -1 */
+        int get_most_repariable_part() const;
```
Yet another improvement could be making these functions return pointers:
```c+
const vehicle_part *get_most_repariable_part() const;
```
This way you can:
- Get rid of the magic numbers (`-1`). Comparison with `nullptr` is clearer.
- Replace the cumbersome `veh->parts[most_repariable].whatever` with the nicer `most_repariable->whatever`.

codemime commented on src/veh_interact.cpp:
```diff
@@ -1224,6 +1227,42 @@ bool veh_interact::overview( std::function<bool(const vehicle_part &pt)> enable,
+    int high_index = -1;
+    int high_damage = 0;
+
+    for(size_t i=0; i< veh->parts.size(); i++){
```
Same here: the function could become cleaner with pointers instead of indexes.

codemime commented on src/veh_interact.cpp:
```diff
@@ -1787,12 +1826,21 @@ void veh_interact::display_stats()
-        const auto iw = utf8_width(_("Most damaged:")) + 1;
-        x[6] += iw;
-        w[6] -= iw;
+        if( mostDamagedPart == most_repariable){
```
Please don't intermix underscores and camelCase. The former should be used.

codemime commented on src/veh_interact.cpp:
```diff
@@ -1787,12 +1826,21 @@ void veh_interact::display_stats()
-        w[6] -= iw;
+        if( mostDamagedPart == most_repariable){
+            mvwprintz(w_stats, y[6], x[6], c_ltgray, _("Most damaged:"));
+            auto iw = utf8_width(_("Most damaged:")) + 1;
```
The string shouldn't be duplicated. A temp. variable could help here.
Same below.

codemime commented on src/veh_interact.cpp:
```diff
@@ -1802,19 +1850,33 @@ void veh_interact::display_stats()
+        // hoff == 1 is already implied and expected - one line is consumed at least.
+        for( size_t i = 8; i < sizeof(y) / sizeof(y[0]); ++i) {
+            y[i] += hoff - 1;
+        }
```
This snippet looks suspiciously similar to the one above. Could be slightly de-duplicated.

codemime commented:
Please format the new code in accordance with the [code style](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/CODE_STYLE.txt).
After fixing the above and applying the style, this PR will be good to go.

zwparchman commented:
Third time's a charm right.

illi-kun commented on src/veh_interact.cpp:
```diff
@@ -860,11 +860,15 @@ bool veh_interact::do_repair( std::string &msg )
+            {
+                vehicle_part *mostRepariable = get_most_repariable_part();
+                if( mostRepariable )
+                {
```
minor note: style regression, '{' should be moved back to 'if' line.

illi-kun commented on src/veh_type.h:
```diff
@@ -163,6 +163,9 @@ class vpart_info
         /** Requirements for repair of this component (per level of damage) */
         requirement_data repair_requirements() const;
 
+        /** Returns weather or not the part is repairable  */
```
"weather"? this looks weird for me.

zwparchman commented:
The fixes to illi-kun's points are up.

codemime commented on src/veh_interact.cpp:
```diff
@@ -1224,6 +1227,50 @@ bool veh_interact::overview( std::function<bool(const vehicle_part &pt)> enable,
+    vehicle_part *high_part = nullptr;
+    int high_damage = 0;
+
+    for( size_t i = 0; i < veh->parts.size(); i++ ) {
```
Can get rid of the indexes completely using a for-range loop: `for( const auto &elem : veh->parts ) {`

codemime commented on src/veh_interact.cpp:
```diff
@@ -1224,6 +1227,50 @@ bool veh_interact::overview( std::function<bool(const vehicle_part &pt)> enable,
+        int dmg = part->base.damage();
+
+        if( dmg > high_damage ) {
+            high_part = part;
```
This won't work very well: `high_damage` is never updated.
Actually, you can get rid of both `dmg` and `high_damage`. I'd also rename `high_part` to `result` or `res`:
```c++
if( res == nullptr || res.damage() < elem.damage() ) {
    res = &elem;
}
```

codemime commented on src/veh_interact.cpp:
```diff
@@ -860,11 +860,14 @@ bool veh_interact::do_repair( std::string &msg )
-                msg = _( "There are no damaged parts on this vehicle." );
-                return false;
+            {
+                vehicle_part *mostRepariable = get_most_repariable_part();
```
Please don't use CamelCase: `const vehicle_part *most_repariable = ...`.

zwparchman commented:
The changes are up
